### DISCUSSION TITLE
Number of DIB checks per cycle

### DISCUSSION THREAD

**Discussion by 0x161e-swei** [ChampSim/ChampSim#448] [Q&A]
Currently [O3_CPU::check_dib](https://github.com/ChampSim/ChampSim/blob/master/src/ooo_cpu.cc#L204) performs the dib check for the next `FETCH_WIDTH` instructions.   
Should we perform only one address lookup per cycle? Meaning that instructions that fall onto the next "window" should not be checked this cycle?

**Comment by ngober** [comment]
Any answer to a "should we" question is going to be fraught with caveats. This is the nature of simulations: they are approximations of the physical structures they model. We certainly could do it differently. I'm interested in making ChampSim a better approximation of physical reality. Do you have reason to believe doing a different DIB lookup would be more correct?

**Reply by 0x161e-swei** [reply]
The `FETCH_WIDTH` number of instructions in `IFETCH_BUFFER` may span across two different windows (or sets) in DIB.
And if these instructions all hit in the DIB, then they are all marked as ready to be moved to the DISPATCH_BUFFER.

A perhaps better model would be only delivering the "micro-ops" within the same micro-op cache entry. In other words, only invoke `do_check_dib` for instructions that are within the same window (set).

**Comment by alberto-ros** [comment]
Let me add something here. In my opinion DIB should have a config parameter called "max_tag_check" (as caches have), and perhaps also a "max_fill" parameter. That is, number of read and write ports. That way the lookup should be done for as many entries (windows) as "max_tag_check" defined.

For the L1I it works like that, and the code should be very similar, so a quick fix in my opinion.

**Comment by ngober** [comment]
What about something like this? https://github.com/ChampSim/ChampSim/commit/788c0acd3817a546d824c649f8978fb2a5f7f326

It's not tested, but it integrates some the things that @alberto-ros is talking about.

**Comment by alberto-ros** [comment]
I checked it, and I see that the dib width represents the number of read ports, and the used ports are not incremented if matching the set and the tag for consecutive instructions, so, yes, it is in the line of what I said (I did not do a code proof though).

**Reply by ngober** [reply]
Right, I would want to do some tests before merging this, but I think this makes a good draft.

**Comment by krishnamtibrewala** [comment]
It would be also important to eventually make DIB inclusive of L1I cache and I-TLB

**Reply by ngober** [reply]
I'll continue pointing people to #350 until my fingers fall off. Help is needed to implement this.
